### PR TITLE
feat(core): Hint support 12 directions

### LINKED ISSUE #699: 🚀 - Allow LineClamp hint to expand to the right
Allow `LineClamp` hint to expand to the right for one line by using `minWidth` instead of `width` and setting width to `min-content` and setting `white-space: nowrap` on hint content

### LINKED ISSUE #739: 🚀 - `Hint` - support 12 directional positioning
# 🐞 Bug report

### Description

Right now [TuiDirection](https://github.com/TinkoffCreditSystems/taiga-ui/blob/173213fa013c73ea38e2afbc16c5ba3f3996bae3/projects/core/types/direction.ts#L7) is declared the following way:
```ts
export type TuiDirection = TuiHorizontalDirection | TuiSideDirection;
```
but I would expect it to also include `TuiVerticalDirection` as well:
```ts
export type TuiDirection = TuiHorizontalDirection | TuiVerticalDirection | TuiSideDirection;
```

### LINKED ISSUE #1258: 🚀 - Customizable Tooltips
### Which @taiga-ui/* package(s) are relevant/releated to the feature request?

core

### Description

Allow css classes to be added to individual tooltips. 

This will allow color changes as well as changing the styles of tooltips as needed.

### PR DESCRIPTION
**Pull request by waterplea**
## PR Checklist

Please check if your PR fulfills the following requirements:

- [ ] The commit message follows [Conventional Commits](https://www.conventionalcommits.org/en/)
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

## PR Type

What kind of change does this PR introduce?

- [ ] Bugfix
- [ ] Feature
- [ ] Refactoring
- [ ] Code style update
- [ ] Build or CI related changes
- [ ] Documentation content changes

## What is the current behavior?

Closes #699
Closes #739
Closes #1258 

## Does this PR introduce a breaking change?

- [x] Yes
- [ ] No

### CODE REVIEW

File: projects/addon-charts/components/line-days-chart/line-days-chart-hint.directive.ts [hunk 1]
```diff
@@ -1,42 +1,37 @@
...
     Directive,
-    ElementRef,
     forwardRef,
     Inject,
     Input,
     NgZone,
     QueryList,
 } from '@angular/core';
-import {ANIMATION_FRAME} from '@ng-web-apis/common';
+import {tuiLineChartDrivers} from '@taiga-ui/addon-charts/components';
```
Comment (nsbarsukov) [inline projects/addon-charts/components/line-days-chart/line-days-chart-hint.directive.ts:11]: probably, it breaks `build`-job

**Review by splincode — APPROVED** [review]

File: projects/core/directives/hint/hint.component.ts [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+        if (driver instanceof TuiHintHoverDirective) {
+            hovered$.pipe(takeUntil(destroy$)).subscribe(hover => driver.toggle(hover));
+        }
+    }
+
+    get content(): PolymorpheusContent {
+        return this.context.$implicit.content;
+    }
+
+    private update(top: number, left: number): void {
```
Comment (vladimirpotekhin) [inline projects/core/directives/hint/hint.component.ts:78]: let's add tuiPure to reduce style recalculations

File: projects/core/directives/hint/hint-position.directive.ts [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+    TuiPositionAccessor,
+    TuiRectAccessor,
+} from '@taiga-ui/core/abstract';
+import {TUI_HINT_DIRECTIONS} from '@taiga-ui/core/constants';
+import {TuiHintDirection, TuiPoint} from '@taiga-ui/core/types';
+
+import {TUI_HINT_OPTIONS, TuiHintOptions} from './hint-options';
+
+const OFFSET = 8;
+const ARROW_OFFSET = 22;
```
[the remark this answers is not in the document]
Comment (waterplea) [reply projects/core/directives/hint/hint-position.directive.ts:15]: One can completely replace implementation if customization is required (like in `LineClamp`). I tried reasonable REM values like 10-25 and it looks acceptable. Not sure we need to overengineer it to account for extreme values.

File: projects/core/directives/hint/hint.component.ts (hunk 2 above)
Comment (waterplea) [reply projects/core/directives/hint/hint.component.ts:78]: It is not pure, it depends on ClientRect values we get in the method.

Comment (vladimirpotekhin) [reply projects/core/directives/hint/hint.component.ts:78]: hmm, got it
but for what cases we need to update ClientRect values without updated (top, left) ? I tried to add pure and didn't find that something breaks

Comment (waterplea) [reply projects/core/directives/hint/hint.component.ts:78]: You are right, we only need these `ClientRect` values to position the beak and for the same `top` and `left` it results in the same position. Nice catch, added `@tuiPure`.

**Review by vladimirpotekhin — APPROVED** [review]

File: projects/core/directives/hint/hint-driver.directive.ts [hunk 4]
```diff
@@ -0,0 +1,9 @@
+import {Directive} from '@angular/core';
+import {TuiDestroyService} from '@taiga-ui/cdk';
+import {AbstractTuiDriverDirective} from '@taiga-ui/core/abstract';
+
+@Directive({
+    selector: `[tuiHint]`,
+    providers: [TuiDestroyService],
+})
+export class TuiHintDriverDirective extends AbstractTuiDriverDirective {}
```
Comment (nsbarsukov) [inline projects/core/directives/hint/hint-driver.directive.ts:9]: Can we explicitly extends this class by all "drivers" ?

It was difficult for me to realize when we subscribe to all "drivers"-observables :(
Now I understand that it happens here but it not obvious.

File: projects/demo/src/modules/app/pages.ts [hunk 5]
```diff
@@ -368,6 +368,36 @@ export const pages: TuiDocPages = [
...
+            },
+            {
+                section: $localize`Tools`,
+                title: `ManualHint`,
+                keywords: `tooltip, тултип, hint, подсказка, помощь, help, manual, программный, хинт`,
+                route: `/directives/hint-manual`,
+            },
+            {
+                section: $localize`Tools`,
+                title: `PointerHint`,
```
Comment (nsbarsukov) [inline projects/demo/src/modules/app/pages.ts:395]:
```suggestion
                title: `HintPointer`,
```

File: projects/demo/src/modules/app/pages.ts [hunk 6]
```diff
@@ -368,6 +368,36 @@ export const pages: TuiDocPages = [
...
+            },
+            {
+                section: $localize`Tools`,
+                title: `Hint`,
+                keywords: `tooltip, тултип, hint, подсказка, помощь, help, хинт`,
+                route: `/directives/hint`,
+            },
+            {
+                section: $localize`Tools`,
+                title: `ManualHint`,
```
Comment (nsbarsukov) [inline projects/demo/src/modules/app/pages.ts:389]:
```suggestion
                title: `HintManual`,
```

File: projects/demo/src/modules/components/abstract/inherited-documentation/inherited-documentation.template.html [hunk 7]
```diff
@@ -132,11 +132,11 @@ <h6 class="tui-text_h6">
         </ng-template>
         <ng-template
             i18n
-            documentationPropertyName="tuiHintMode"
+            documentationPropertyName="tuiHintAppearance"
```
Comment (nsbarsukov) [inline projects/demo/src/modules/components/abstract/inherited-documentation/inherited-documentation.template.html:135]: Also,

`documentationPropertyType="TuiDirection"` => `documentationPropertyType="TuiHintDirection"`

(8 lines above)

File: projects/demo/src/modules/directives/hint-controller/hint-controller.template.html [hunk 8]
```diff
@@ -52,11 +52,11 @@
...
                 i18n
-                documentationPropertyName="tuiHintMode"
+                documentationPropertyName="tuiHintAppearance"
                 documentationPropertyMode="input"
-                documentationPropertyType="TuiTooltipMode | null"
-                [documentationPropertyValues]="modeVariants"
-                [(documentationPropertyValue)]="tuiHintMode"
+                documentationPropertyType="string"
+                [documentationPropertyValues]="appearanceVariants"
+                [(documentationPropertyValue)]="tuiHintAppearance"
```
Comment (nsbarsukov) [inline projects/demo/src/modules/directives/hint-controller/hint-controller.template.html:59]: `TuiDirection` => `TuiHintDirection`

**Review by nsbarsukov — APPROVED** [review]

File: projects/core/directives/hint/hint-driver.directive.ts (hunk 4 above)
Comment (waterplea) [reply projects/core/directives/hint/hint-driver.directive.ts:9]: The idea is we have a driver and we have a vehicle. To connect them together we need to inject and subscribe. And to initiate this action we need a directive with a particular selector. But the action is always the same so I moved it into abstract directive. Basically, for every pair of driver/vehicle we will have a directive like that which is only necessary for its selector and the work it does is the same and happens in abstract class.

**Pull request merged**